### ISSUE TITLE
False positive cast diagnostics with DSTs and `zerocopy`

### ISSUE THREAD

**Issue by ShoyuVanilla** [rust-lang/rust-analyzer#19431]
> There's still a problem with `KnownLayout` when the DST is wrapped:
> 
> ```rust
> use zerocopy::KnownLayout;
> 
> #[derive(KnownLayout)]
> #[repr(C)]
> pub struct Dst([u8]);
> 
> #[derive(KnownLayout)]
> #[repr(C)]
> pub struct Struct {
>     pub dst: Dst,
> }
> ```
> 
> ```
> cannot cast thin pointer `*mut __ZerocopyKnownLayoutMaybeUninit` to fat pointer `*mut _` rust-analyzer[E0607]
> ```
> 
> `rust-analyzer 1.85.1 (4eb16125 2025-03-15)`
> 
> No problem with rustc. 

 _Originally posted by @cloneable in [#18682](https://github.com/rust-lang/rust-analyzer/issues/18682#issuecomment-2746185008)_

**Comment by cloneable** [comment]
Thank you!
